### PR TITLE
Parameter Spec

### LINKED ISSUE #86: Parameter Spec
Update the `model.json` to include a parameters section for each model (not deprecated) with a `parameters` field specifying for each parameter:
- `default` 
- `min` (only for `max_tokens` and `temperature`)
- `max` (only for `max_tokens` and `temperature`)

> Note
> Specify default value for each parameter specified in the chat completion (e.g [openai](https://github.com/premAI-io/prem-utils/blob/main/prem_utils/connectors/openai.py#L74-L88))
> parameters["temperature"] = model_parameters.temperature
> parameters["top_p"] = model_parameters.top_p
> parameters["max_tokens"] = model_parameters.max_tokens
> parameters["presence_penalty"] = model_parameters.presence_penalty
> parameters["frequency_penalty"] = model_parameters.frequency_penalty
> parameters["logit_bias"] = model_parameters.logit_bias

### PR DESCRIPTION
**Pull request by vittoriop17**
closes #86

### CODE REVIEW

**Review by nsosio — COMMENTED** [review]
Before checking it, @vittoriop17 how have you set the default parameters? Because some of them are at the limit (either min or max) and other not

**Comment by vittoriop17** [comment]
> how have you set the default parameters? 

@nsosio good point. I will go through each parameter:
- temperature: I always set it to (max-min)/2. To avoid too random responses. But also to avoid too deterministic ones. Regarding this argument, please note that for perplexity models, max_temperature=1.99. I had to do this, since pplx temperature range is [0.0, 2), different from other providers that have [0.0, 2.0] or [0.0, 1.0];
- frequency_penalty and presence_penalty: they range in [-2.0, 2.0] for almost all the models (apart from a few exceptions). I set the default to 0, for all the models that range between -2 and 2. Most of the providers set 0.0 as default, and so do I. Regarding the exceptions: cohere models range between [0.0, 1.0], and the default is set to 0.0 (like in their [docs](https://docs.cohere.com/reference/chat)). Instead, for perplexity models, the [frequency_penalty ](https://docs.perplexity.ai/reference/post_chat_completions)must be greater than 0, and a value of 1.0 means no penalty.
- top_p: usually ranges in [0.0, 1.0]. I set it to 0.5 for almost every model. I set it to 0.0 for a few of them. **But I'm actually considering to remove it from all the models**. The reason? In every documentation, it is suggested to tune temperature OR top_p, not both at the same time. It is recommended for advanced use cases only.
- max_tokens: I should make more checks on this. But the idea is to set it to the max value. Nevertheless, there could be some problems at runtime. In some docs, I found the following:  `max_tokens + input_tokens <= context_length`. This could be a problem if the current max_tokens_default is set to the maximum (that sometimes matches with the context_length). When invoking the chat_completions, it will raise an error. That's why for models like azure-mistral or azure-gpt I set the default to max/2 or max/4 (because they do the check I reported above)

**Comment by nsosio** [comment]
> > how have you set the default parameters?
> 
> @nsosio good point. I will go through each parameter:
> 
> * temperature: I always set it to (max-min)/2. To avoid too random responses. But also to avoid too deterministic ones. Regarding this argument, please note that for perplexity models, max_temperature=1.99. I had to do this, since pplx temperature range is [0.0, 2), different from other providers that have [0.0, 2.0] or [0.0, 1.0];
> * frequency_penalty and presence_penalty: they range in [-2.0, 2.0] for almost all the models (apart from a few exceptions). I set the default to 0, for all the models that range between -2 and 2. Most of the providers set 0.0 as default, and so do I. Regarding the exceptions: cohere models range between [0.0, 1.0], and the default is set to 0.0 (like in their [docs](https://docs.cohere.com/reference/chat)). Instead, for perplexity models, the [frequency_penalty ](https://docs.perplexity.ai/reference/post_chat_completions)must be greater than 0, and a value of 1.0 means no penalty.
> * top_p: usually ranges in [0.0, 1.0]. I set it to 0.5 for almost every model. I set it to 0.0 for a few of them. **But I'm actually considering to remove it from all the models**. The reason? In every documentation, it is suggested to tune temperature OR top_p, not both at the same time. It is recommended for advanced use cases only.
> * max_tokens: I should make more checks on this. But the idea is to set it to the max value. Nevertheless, there could be some problems at runtime. In some docs, I found the following:  `max_tokens + input_tokens <= context_length`. This could be a problem if the current max_tokens_default is set to the maximum (that sometimes matches with the context_length). When invoking the chat_completions, it will raise an error. That's why for models like azure-mistral or azure-gpt I set the default to max/2 or max/4 (because they do the check I reported above)

1. ok
2. ok
3. let's remove it
4. I would like to understand for each model if max_tokens is equal to output_tokens or the sum of input and output but let's handle it in a separate issue because we're going through a refractory. As of know I would set the default always to the max value

**Comment by vittoriop17** [comment]
> I would like to understand for each model if max_tokens is equal to output_tokens or the sum of input and output

@nsosio I know it would be better to handle it in a separate issue, but it will cause some errors at runtime. Every time a model provider checks if `max_tokens + input_tokens <= context-window`, there will be an exception like this:

`Error for model gpt-4-32k-azure: {'message': 'Error code: 400 - {\'error\': {\'message\': "**_This model\'s maximum context length is 32768 tokens. However, you requested 32787 tokens (19 in the messages, 32768 in the completion_**). Please reduce the length of the messages or completion.", \'type\': \'invalid_request_error\', \'param\': \'messages\', \'code\': \'context_length_exceeded\'}}', 'code': 'UnknownError', 'status_code': 500, 'document_chunks': [], 'trace_id': 'ead5b613-4a13-4303-bec8-545fe2290e3a'} `

then, if I set max_tokens to max_value, the chat_completion request won't be completed.
There could be two possible solutions:
- add a boolean flag that identifies all the models that pass through this check. In this case, we can set the max_tokens = max_value/2.
- remove the default max_tokens parameter from all the models but claude ones. Looking at the docs, I found that only for Anthropic APIs max_tokens argument is mandatory. Thus, for other providers' APIs, we can not pass it, and the provider will automatically handle that

File: prem_utils/models.json [hunk 1]
```diff
@@ -189,7 +460,35 @@
...
+                    {
+                        "temperature": {
+                            "default": 1.0,
+                            "min": 0.0,
+                            "max": 2.0
+                        },
+                        "max_tokens": {
+                            "default": 4096,
+                            "min": 0,
+                            "max": 4096
```
Comment (nsosio) [inline prem_utils/models.json:474]: why max_tokens less than context window?

Comment (vittoriop17) [reply prem_utils/models.json:474]: I tested gpt-4-turbo-preview with prem-utils connector and also with openAI chat API. And I figured out that 4096 is the upper limit. Although, openAI [docs ](https://platform.openai.com/docs/api-reference/chat/create) are not super clear on this

**Review by nsosio — APPROVED** [review]

**Pull request merged**